### PR TITLE
feat: add active days endpoint under v1-alpha

### PR DESCRIPTION
**Pull request by epipav**
## What

Adds `GET /v1-alpha/projects/{slug}/development/active-days`, the API port of the Active Days widget handler (`frontend/server/api/widget/development/active-days.get.ts`).

Query: `repos[]`, `startDate`, `endDate` (UTC calendar days), `granularity` (required: daily, weekly, monthly, quarterly, yearly) and `includeCollaborations` (optional boolean, default `false`).

Response:

```json
{
  "summary": { "current": 80, "previous": 64, "percentageChange": 25, "changeValue": 16,
               "periodFrom": "2025-01-01T00:00:00Z", "periodTo": "2025-03-31T00:00:00Z" },
  "avgContributionsPerDay": 21.5,
  "data": [
    { "startDate": "2025-01-01T00:00:00Z", "endDate": "2025-01-31T00:00:00Z", "contributions": 698 }
  ]
}
```

`summary` counts active days (days with at least one development activity) in the current period against the previous one, which has the same length and ends the day before `startDate`. `data` has one bucket per granularity step with the contributions in it.

Three concurrent calls to the `active_days` pipe (current summary, previous summary, series), as the Nuxt data layer makes them, with the slug sent as `project`.

## Changes vs the Nuxt handler

- `day` (the 1-based bucket index) is dropped.
- `avgContributions` is renamed to `avgContributionsPerDay`, the pipe's own name.
- An empty Tinybird result returns 200 with zeros and `data: []`. Nuxt reads `currentSummary.data[0]` unguarded and returns a 500.
- `percentageChange` is signed (Nuxt used `Math.abs`).
- Query dates are `YYYY-MM-DD`; response dates are ISO-8601 UTC; Tinybird gets `YYYY-MM-DD 00:00:00` exactly as Nuxt's Luxon serializer sends it.

## Decisions

- **`includeCollaborations` is exposed** (optional boolean, default `false`) and passed to the pipe. The Nuxt handler hardcodes `false`, which is why the widget's collaboration toggle does nothing. The pipe accepts the flag: Nuxt's series call already sends `includeCodeContributions=true&includeCollaborations=false` (it spreads the whole filter), and `ActiveDaysFilter` declares both flags. Default `false` keeps the default response identical to today's widget. Same param as IN-1335.
- **The flags go on all three pipe calls**, summaries included. Nuxt sends them only on the series call, so with the flag exposed the summary would disagree with the series whenever `includeCollaborations=true`. This follows the `active_contributors` data layer, which sends both flags on all three of its calls.
- **`includeCodeContributions` stays hardcoded to `true`**, as in Nuxt. Exposing it later is additive.
- **`granularity` is required.** Nuxt relies on a pipe default that is not visible from this repo; a required param keeps the series shape explicit instead of documenting a default we cannot verify.
- **An unknown slug returns 200 with zeros** (epic definition of done for metric endpoints). The Tinybird client throws `TinybirdProjectNotFoundError` when the bucket lookup finds no row for the slug; the route catches exactly that class and treats the call as an empty result. A 404 from the pipe itself still maps to 503, and a test pins that.
- **A null `avgContributionsPerDay` becomes `0`.** The column is `Nullable(Float64)` and is null only with zero active days, which the empty case already reports as `0`.
- **A bare `repos=` is dropped** rather than sent to the pipe as a filter on an empty repository name (Ajv coerces it to `['']`), matching the Nuxt handler.
- Every Tinybird failure (any status, network error, non-JSON body) maps to 503 `upstream_unavailable`, logged once per request. A missing `API_TB_*` variable stays a 500, as on `/projects/{slug}`. Cache-Control is inherited from the alpha plugin. OpenAPI tag `Development`; every response field and added query param carries a description with its unit.

## Tests

`api/tests/development-active-days.test.ts`: 32 cases through `buildApp()` with `fetch` stubbed at the HTTP boundary, so the real Tinybird client builds the URLs, resolves the bucket and classifies the errors. Covers the response shape and key set, the exact params of the three pipe calls and their concurrency, the `includeCollaborations` pass-through and validation, empty `repos`, signed and null `percentageChange`, empty results, unknown slug, date and granularity validation, the 503 mapping (pipe 500/404/401/429, bucket lookup 500, network error, non-JSON), Cache-Control, and the OpenAPI spec (tag, field descriptions, query params, absent from `/v1`).

From `api/`: `pnpm test` (14 files, 194 tests), `pnpm tsc-check`, `pnpm lint`, `pnpm format:check` all pass. No new runtime dependency.

Includes the routing-test parser fix from the base branch (62b136b9).

Local helpers to fold into a shared module once the five Development PRs land: `toTinybirdRange`, `toUtcDateTime` (same form as `period.ts`'s private `utcMidnight`), the `fetchRows` wrapper (bucket-miss to empty rows) with the `.catch` 503 mapping, and the `Type.Composite([DateRangeQuery, ...])` query schema. The api's Tinybird client has no bucket cache, so each request makes three `project_buckets` lookups for the same slug; wiring the client's in-process cache fixes that for every route.

Based on main, which includes #2241 (IN-1348, directory-based route registration) as dd37a7a9. Jira: IN-1334

- `main` 
  - \#2241
    - **feat: add active days endpoint under v1-alpha** :point\_left:

### CODE REVIEW

File: api/src/versions/v1-alpha/development/active-days.ts [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    async (request) => {
+      const { slug } = request.params;
+      const {
+        repos,
+        startDate,
+        endDate,
+        granularity,
+        includeCollaborations = false,
+      } = request.query;
+      const dates = getPreviousDates(startDate, endDate);
```
[the remark this answers is not in the document]
Comment (epipav) [reply api/src/versions/v1-alpha/development/active-days.ts:138]: Fixed in 1a8ae905: description now says calendar span (months plus days) instead of same length. `getPreviousDates` itself stays, it mirrors the UI's luxon diff so both agree day for day.

File: api/tests/development-active-days.test.ts [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+    const res = await get(url({ startDate: undefined, endDate: undefined }));
+    expect(res.statusCode).toBe(200);
+    expect(res.json().summary).toMatchObject({
+      periodFrom: isoDay('2010-01-01'),
+      periodTo: isoDay(today),
+    });
+    const series = pipeCalls().find((call) => call.searchParams.has('granularity'));
+    expect(series?.searchParams.get('startDate')).toBe(atMidnight('2010-01-01'));
+    expect(series?.searchParams.get('endDate')).toBe(atMidnight(today));
+  });
```
[the remark this answers is not in the document]
Comment (epipav) [reply api/tests/development-active-days.test.ts:447]: Fixed in 1a8ae905: the test fakes only `Date` and restores real timers in a `finally`, so the client's timers keep running.

File: api/src/versions/v1-alpha/development/active-days.ts [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    description: 'Last calendar day of the bucket, at 00:00:00 UTC.',
+  }),
+  contributions: Type.Integer({ description: 'Number of contributions in the bucket (count).' }),
+});
+
+const ActiveDays = Type.Object({
+  summary: {
+    ...PeriodSummary,
+    description:
+      'Active days in the current period against the previous one. current, previous and changeValue are counts of days.',
```
[the remark this answers is not in the document]
Comment (epipav) [reply api/src/versions/v1-alpha/development/active-days.ts]: Fixed in 1a8ae905: summary schema is built from `PeriodSummary.properties` with the three counts described in days, test checks all six children. `schemas/common.ts` is shared by five parallel PRs, so its own descriptions land in the consolidation chore.

**Review by gaspergrom — APPROVED** [review]
Checked the includeCollaborations fan-out across all three pipe calls and the TinybirdProjectNotFoundError catch, both hold up, and the null avgContributionsPerDay path lands on 0. Tests cover the edge cases well. LGTM.

File: api/src/versions/v1-alpha/development/active-days.ts [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+      return {
+        summary: toPeriodSummary(
+          current?.activeDaysCount ?? 0,
+          previous?.activeDaysCount ?? 0,
+          dates.current,
+        ),
+        avgContributionsPerDay: current?.avgContributionsPerDay ?? 0,
+        data: seriesRows.map((row) => ({
+          startDate: toUtcDateTime(row.startDate),
+          endDate: toUtcDateTime(row.endDate),
```
Comment (themarolt) [inline api/src/versions/v1-alpha/development/active-days.ts]: the pipe meta in `frontend/server/mocks/tinybird-active-days-response.mock.ts` types the series `startDate`/`endDate` as `Nullable(Date)` - a null there interpolates into `"nullT00:00:00Z"` and goes out as a `date-time` field, and fast-json-stringify won't catch it since it's already a string. do those get the same treatment you gave the nullable `avgContributionsPerDay`, or is the column never null on a row the pipe returns?

**Review by themarolt — APPROVED** [review]
lgtm - one question on the nullable bucket dates in the series

File: api/src/versions/v1-alpha/development/active-days.ts (hunk 4 above)
Comment (epipav) [reply api/src/versions/v1-alpha/development/active-days.ts]: confirmed, both bounds are `Nullable(Date)` in the meta and the pipe sql isn't in this repo, so the handler guards now. Fixed in 40d1111b: rows missing either bound are dropped before formatting, with a test pinning it :+1:

File: api/src/versions/v1-alpha/development/active-days.ts [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+      const currentRange = toTinybirdRange(dates.current);
+      const previousRange = toTinybirdRange(dates.previous);
+      const activeDays = <T>(query: TinybirdQuery) =>
+        fromTinybird(log, 'active_days', () => client.fetch<T>(pipePath, query));
+
+      const [currentRows, previousRows, seriesRows] = await Promise.all([
+        activeDays<SummaryRow[]>({ ...shared, ...currentRange }),
+        activeDays<SummaryRow[]>({ ...shared, ...previousRange }),
+        activeDays<SeriesRow[]>({ ...shared, ...currentRange, granularity }),
+      ]);
```
[the remark this answers is not in the document]
Comment (epipav) [reply api/src/versions/v1-alpha/development/active-days.ts:176]: each call logs its own failure with its pipe label on purpose, same as `issues-resolution.ts` on main; wrapping the aggregate would hide the other two rejections. the shared wrapper and its logging policy land in IN-1347.
